task 70: String path -> Path path

tieSari opened the pull request:
Nyt olisi kutakuinkin string polut muutettu nio.Path-muotoon. Joihinkin testeihin jätin stringiksi, jos niissä haluttiin polku String[] muotoon polun osien läpikäyntiä varten.

DarthKipsu committed:
changes all url string to URI type

Fixes #77

refactor strings to URI changes

arrange two more imports

fix indentation

fix indentation and import orders

fixed little bug

Brought back zip_url because it got lost after rebasing

DarthKipsu committed:
changes all url string to URI type

Fixes #77

refactor strings to URI changes

arrange two more imports

DarthKipsu committed:
changes all url string to URI type

Fixes #77

refactor strings to URI changes

arrange two more imports

wakr committed:
There is something wrong with the DownloadExercisesTest and course.json zip_url-field

fix indentation

fix indentation and import orders

wakr committed:
There is something wrong with the DownloadExercisesTest and course.json zip_url-field

fix indentation

fix indentation and import orders

Commit:
task 70: String path -> Path path

ljleppan commented on src/main/java/fi/helsinki/cs/tmc/core/commands/DownloadModelSolution.java:
```diff
@@ -41,10 +41,9 @@ public Boolean call() throws Exception {
-                                settings.getTmcMainDirectory(), courseName));
+        Path target
+                = exerciseDownloader.createCourseFolder(
+                        Paths.get(settings.getTmcMainDirectory()), courseName);
```
Will this result in NPE if `settings.getTmcMainDirectory()` returns `null`? 

Checking for `null` every time `settings.getTmcMainDirectory()` is called seems cumbersome. Do we have a nice way of declaring that either the interface must always return a non-null value or otherwise doing this check in one place only?

tieSari replied:
Perhaps TmcMainDirectory could also be Path, not String?

ljleppan commented on src/main/java/fi/helsinki/cs/tmc/core/commands/RunCheckStyle.java:
```diff
@@ -4,6 +4,7 @@
 import fi.helsinki.cs.tmc.langs.domain.NoLanguagePluginFoundException;
 import fi.helsinki.cs.tmc.langs.util.TaskExecutorImpl;
+import java.nio.file.Path;
 
```
Import order: this should be below the empty line.

See https://google.github.io/styleguide/javaguide.html#s3.3.3-import-ordering-and-spacing, replacing `com.google` with `fi.helsinki`.

ljleppan commented on src/main/java/fi/helsinki/cs/tmc/core/commands/RunTests.java:
```diff
@@ -6,6 +6,7 @@
 import fi.helsinki.cs.tmc.langs.util.TaskExecutor;
 import fi.helsinki.cs.tmc.langs.util.TaskExecutorImpl;
+import java.nio.file.Path;
 
```
Import order, see above

ljleppan commented on src/main/java/fi/helsinki/cs/tmc/core/communication/ExerciseDownloader.java:
```diff
@@ -100,14 +100,14 @@ public ExerciseDownloader(UrlCommunicator urlCommunicator, TmcApi tmcApi) {
-            coursePath.mkdirs();
+
+        if (!Files.exists(path)) {
+            path.toFile().mkdirs();
```
Prefer [`java.nio.Files#createDirectories(Path dir, FileAttribute<?>... attrs)`](http://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#createDirectories%28java.nio.file.Path,%20java.nio.file.attribute.FileAttribute...%29)  (and maybe [`java.nio.Files#notExists(Path path, LinkOption... options)`](http://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#notExists%28java.nio.file.Path,%20java.nio.file.LinkOption...%29) on the line above, although this has slightly different semantics)

ljleppan commented on src/main/java/fi/helsinki/cs/tmc/core/communication/ExerciseDownloader.java:
```diff
@@ -194,8 +179,8 @@ public double getPercents(int exCount, int exercisesSize) {
-    private void downloadExerciseZip(String zipUrl, String path) {
-        File file = new File(path);
+    private void downloadExerciseZip(String zipUrl, Path path) {
+        File file = new File(path.toString());
```
Is it possible to replace this `File` with `Path` as well? It's a bit out of scope of this issue so if it's non-trivial we can generate a new issue for this migration.

ljleppan commented on src/test/java/fi/helsinki/cs/tmc/core/testhelpers/FileWriterHelper.java:
```diff
@@ -15,9 +16,9 @@
+    public void writeStuffToFile(Path path) throws IOException {
         try {
-            writer = new FileWriter(new File(path));
+            writer = new FileWriter(new File(path.toString()));
```
`path.toFile()` should be simpler.

jamo commented on src/test/java/fi/helsinki/cs/tmc/core/testhelpers/ProjectRootFinderStub.java:
```diff
@@ -14,12 +14,12 @@
-    private String returnValue;
+    private Path returnValue;
     private TmcApi tmcApi;
     private HashMap<String, Course> courseStubs;
```
Outside of the scope of this PR, but s/HashMap/Map/

ljleppan commented on src/test/java/fi/helsinki/cs/tmc/core/commands/RunTestsTest.java:
```diff
@@ -5,6 +5,8 @@
 import fi.helsinki.cs.tmc.core.exceptions.TmcCoreException;
 import fi.helsinki.cs.tmc.langs.domain.RunResult;
+import java.nio.file.Path;
+import java.nio.file.Paths;
```
Import order

ljleppan commented on src/test/java/fi/helsinki/cs/tmc/core/commands/PasteWithCommentTest.java:
```diff
@@ -21,7 +21,10 @@
+import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.text.ParseException;
+import static org.mockito.Matchers.any;
```
Import order

ljleppan commented on src/test/java/fi/helsinki/cs/tmc/core/commands/DownloadModelSolutionTest.java:
```diff
@@ -5,10 +5,12 @@
 import fi.helsinki.cs.tmc.core.communication.TmcApi;
 import fi.helsinki.cs.tmc.core.domain.Course;
 import fi.helsinki.cs.tmc.core.domain.Exercise;
+import java.nio.file.Path;
```
Import order

ljleppan commented on src/test/java/fi/helsinki/cs/tmc/core/commands/SubmitTest.java:
```diff
@@ -45,6 +45,7 @@
 import java.text.ParseException;
 import java.util.ArrayList;
 import java.util.List;
+import static org.mockito.Matchers.any;
```
Import order

ljleppan commented on src/test/java/fi/helsinki/cs/tmc/core/testhelpers/ProjectRootFinderStub.java:
```diff
@@ -33,21 +33,21 @@ private void fillCourseStubs() {
-    public Optional<Course> getCurrentCourse(String path) {
-        String[] folders = path.split("\\" + File.separator);
+    public Optional<Course> getCurrentCourse(Path path) {
+        String[] folders = path.toString().split("\\" + File.separator);
```
Can we use [`java.nio.Path#iterator()`](http://docs.oracle.com/javase/7/docs/api/java/nio/file/Path.html#iterator%28%29) here?

ljleppan commented:
Added some comments, most of which are minor details about import ordering. Aside from those, good work!

Before we can merge this PR, you'll need to rebase your work on top of the current `master` branch. If you are unsure on how to do this, ask in IRC and someone will be able to help.

I also noticed that you added some completely new files that seem to be build logs etc.: these should be removed before we can merge the PR.

jamo commented:
Nice work, not much to add , but why are we adding the new test resource exercise testResources/viikko1/Viikko1_001.Nimi

tieSari commented:
Ok, I will fix those things you mentioned. I have no idea where those extra files came from, sorry, I will remove them.

wakr committed:
course.json's urls are now changed to example.com which will be used later as a placeholder for example Wiremock's localhost.
DownloadExercisesTest also modified for this.

2/5 of .jsons now fixed.

4/5 of .jsons now mainly uses example.com which is later(dynamically) changed to 127.0.0.1
if wiremock is present while testing

All .jsons and tests are now free of the "tmc.mooc.fi"-address. Also other
real addresses were changed to example.com.

Fixed little bug where 153 was 152. Correct amount of courses in course.json is 153 (checked with grep)

wakr committed:
course.json's urls are now changed to example.com which will be used later as a placeholder for example Wiremock's localhost.
DownloadExercisesTest also modified for this.

2/5 of .jsons now fixed.

4/5 of .jsons now mainly uses example.com which is later(dynamically) changed to 127.0.0.1
if wiremock is present while testing

All .jsons and tests are now free of the "tmc.mooc.fi"-address. Also other
real addresses were changed to example.com.

Fixed little bug where 153 was 152. Correct amount of courses in course.json is 153 (checked with grep)

Commit:
Replaced generated urls with urls from courses.json

Commit:
Replaced Optional.absent() with Optional.fromNullable(T)

Commit:
task 70 fixies

Commit:
Merge github.com:testmycode/tmc-core

Conflicts:
	src/main/java/fi/helsinki/cs/tmc/core/communication/ExerciseDownloader.java
	src/main/java/fi/helsinki/cs/tmc/core/communication/ExerciseSubmitter.java
	src/main/java/fi/helsinki/cs/tmc/core/communication/UrlCommunicator.java
	src/test/java/fi/helsinki/cs/tmc/core/commands/SubmitTest.java

Commit:
removed extra files

tieSari commented:
Now I have fixed those issues you asked. I tried to make "rebase" after merge an push but apparently it was too late as all files was already up-to-date then.

wakr committed:
Fixed huge spam from pom.xml which happened because
checkstyle failed to locate the correct checkstyle.xml-file

jamo commented on src/main/java/fi/helsinki/cs/tmc/core/communication/UrlCommunicator.java:
```diff
@@ -213,9 +216,14 @@ public boolean downloadToFile(URI url, File file, String credentials) {
      * Calls downloadToFile with username and password as params.
+     * @param url
+     * @param path
+     * @return 
```
Let's remove these empty @-clauses

jamo commented on src/main/java/fi/helsinki/cs/tmc/core/communication/UrlCommunicator.java:
```diff
@@ -213,9 +216,14 @@ public boolean downloadToFile(URI url, File file, String credentials) {
+
+    public boolean downloadToFile(URI url, Path path) {
+        return downloadToFile(url, path, this.settings.getFormattedUserData());
+
```
Remove blank line

jamo commented on src/main/java/fi/helsinki/cs/tmc/core/configuration/TmcSettings.java:
```diff
@@ -3,6 +3,7 @@
 import fi.helsinki.cs.tmc.core.domain.Course;
 
 import com.google.common.base.Optional;
+import java.nio.file.Path;
```
Import order should be:
- All static imports in a single group
- fi.helsinki.cs.tmc imports (only if this source file is in the fi.helsinki.cs.tmc package space)
- Third-party imports, one group per top-level package, in ASCII sort order
- for example: android, com, junit, org, sun
- java imports
- javax imports

tieSari replied:
Hmm, what is wrong with that order? Missing empty line before java.nio?

ljleppan replied:
Yes, every group (saa @jamox above) should be separated by an empty line.

See [this](https://google.github.io/styleguide/javaguide.html#s3.3.3-import-ordering-and-spacing), but replace `com.google` with `fi.helsinki.cs.tmc`.

jamo commented on src/test/java/fi/helsinki/cs/tmc/core/commands/DownloadExercisesTest.java:
```diff
@@ -118,7 +118,7 @@ public void writesChecksumsToCacheIfCacheFileIsGiven()
         CoreTestSettings settings1 = createSettingsAndWiremock();
         core = new TmcCore(settings1);
-        Path folder = Paths.get(System.getProperty("user.dir") + "/testResources/");
+        Path folder = Paths.get(System.getProperty("user.dir") , "testResources");
```
Is this doing what I think it is -- using a normal directory as temp directory - outside or temp folders...

jamo commented on src/test/java/fi/helsinki/cs/tmc/core/commands/DownloadModelSolutionTest.java:
```diff
@@ -5,10 +5,14 @@
+import static org.mockito.Matchers.any;
 import static org.mockito.Matchers.anyString;
 import static org.mockito.Matchers.eq;
 import static org.mockito.Mockito.mock;
```
Import order should be:
- All static imports in a single group
- fi.helsinki.cs.tmc imports (only if this source file is in the fi.helsinki.cs.tmc package space)
- Third-party imports, one group per top-level package, in ASCII sort order
- for example: android, com, junit, org, sun
- java imports
- javax imports

jamo commented on src/test/java/fi/helsinki/cs/tmc/core/commands/RunTestsTest.java:
```diff
@@ -6,6 +6,9 @@
 
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
```
Import order should be:
- All static imports in a single group
- fi.helsinki.cs.tmc imports (only if this source file is in the fi.helsinki.cs.tmc package space)
- Third-party imports, one group per top-level package, in ASCII sort order
- for example: android, com, junit, org, sun
- java imports
- javax imports

jamo commented:
Added a few nitpicks, you probably need to update 'upstream/master' before rebasing; or maybe you could say `git pull upstream master --rebase`

Commit:
minor fixies to task70

tieSari commented:
Hi!
New try again :) I Fixed those things you asked and rebased my fixies. Can you check if it is ok now?

ljleppan commented:
It seems there's still some sort of a conflict between your branch and our master branch. Most likely you rebased against the wrong branch.

You should read https://git-scm.com/book/en/v2/Git-Branching-Rebasing. In this case, the branches in the first example correspond as `experiment` == `this branch` and `master == testmycode/master`. For this pull request, what's called "a simple rebase" in the above link should suffice. 

If you have trouble after reading the above, ask for help in IRC.

tieSari commented:
this can be removed, #92 is the right one.
